core/src/main/scala/org/http4s/client/jdkhttpclient/WebSocketClient.scala

```diff
@@ -0,0 +1,167 @@
+package org.http4s.client.jdkhttpclient
+
+import cats._
+import cats.data.Chain
+import cats.effect._
+import cats.effect.concurrent.{Deferred, Ref, TryableDeferred}
+import cats.implicits._
+import fs2.{Pipe, Pull, Stream}
+import org.http4s.{Headers, Method, Uri}
+import scodec.bits.ByteVector
+
+/**
+  * A websocket request.
+  *
+  * @param uri The URI.
+  * @param headers The headers to send. Some websocket clients reject protocol-specific headers.
+  * @param method The method of the intial HTTP request. Ignored by some clients.
+  * @param subprotocols The subprotocols, descending from most prefered.
+  */
+case class WSRequest(
+    uri: Uri,
+    headers: Headers = Headers.empty,
+    method: Method = Method.GET,
+    subprotocols: List[String] = List.empty
```

Comment:
I would bet not all `String`s are valid subprotocols.  There might be a missing type here.

Reply:
A better representation would express the subprotocols as a `ListSet`, and replace `String` by a newtype ensuring that it is a value HTTP header value (`field-name` from RFC 7230). I think this adds zero value in practice, as subprotocols are 1. rarely used and 2. almost always static.

If I think about it, it is probably better to remove this from  `WSRequest` and let the user supply a `Sec-WebSocket-Protocol` header on their own.

Reply:
I am confused - http4s `RequestCookie`s are not validated. Am I missing sth?

Reply:
You mean invalid `RequestCookie`s can be created? It might be something we just haven't tightened down from our Spray origins.

I think I agree with the header approach.